---
title: "Assay report"
output: html_document
params: 
  assay_selection: assay_selection
  alignment: alignment
  sequence_profile: sequence_profile 
  sequence_properties: sequence_properties 
   
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, eval = TRUE)
```
<br><br>
The assay was designed from an alignment of `r length(params$alignment)` sequences. The length of the alignment/region of interest was `r nchar(params$alignment[[1]])` nucleotides. The report was generated `r Sys.Date()`.
<br><br>
```{r}
assay_report <- print_assay_report(params$assay_selection)
```

```{r, fig.width=12, fig.height=8}
assay_overview_plot(params$sequence_properties, params$assay_selection)
```
<br><br>
**Fig. 1.** Overview of the alignment/region of interest used for assay design, with the assay region indicated in blue. Identity: Proportion of the most frequently occurring base. Entropy: Shannon entropy, a measurement of variability. A value of zero indicate no variability and a high value indicate high variability. GC: GC-content. Gaps: Proportion of gaps. Vertical bars represent the value at each position. Black bars in the two uppermost figures represent values of 1 and 0, respectively (i.e. positions with complete conservation). The horizontal lines represent centered running averages.   
<br><br>
```{r, fig.width=12, fig.height=6}
assay_detail_plot(params$sequence_profile, params$assay_selection)
```
<br><br>
**Fig. 2.** Proportion of each nucleotide in the alignment at the oligo binding sites. Sequences are displayed from 5'- to 3'-end. 
<br><br>
**Table 1**. Assay info. Begin: Where the assay region starts (with regards to the alignment), End: Where the assay region ends. Tm difference, primers: Tm difference between the two primers (absolute value, in C). Tm difference, probe-primers: The Tm of the probe subtracted by the average Tm of the primer pair. Note that the Tm-differences are calculated from the majority oligos, and may thus be misleading for degenerate (IUPAC) oligos. Degeneracy; total number of oligo variants in the assay.  
<br>
```{r}
if (any(grepl("^pm_", names(assay_report[[1]][[1]])))) {
  knitr::kable(assay_report[[1]]$general, "html", digits = 2, col.names = c("Begin", "End", "Amplicon length", "Tm difference, primers", "Tm difference, probe-primers", "Degeneracy", "Perfect match (majority)", "Perfect match (IUPAC)")) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), font_size = 11, full_width = TRUE)
} else {
  knitr::kable(assay_report[[1]]$general, "html", digits = 2, col.names = c("Begin", "End", "Amplicon length", "Tm difference, primers", "Tm difference, probe-primers", "Degeneracy")) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), font_size = 11, full_width = TRUE)
}
```
<br>
**Table 2.** Forward primer, summary. 
```{r}
if (any(grepl("^pm_", names(assay_report[[1]][[1]])))) {
knitr::kable(assay_report[[1]]$forward, digits = 2, col.names = c("Begin", "End", "Length", "Sequence (majority)", "Sequence (IUPAC)", "Degenerates", "Degeneracy", "GC-content (majority)", "Tm (majority)", "Perfect matches (majority)", "Perfect matches (IUPAC)")) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), font_size = 11, full_width = TRUE)
} else {
knitr::kable(assay_report[[1]]$forward, digits = 2, col.names = c("Begin", "End", "Length", "Sequence (majority)", "Sequence (IUPAC)", "Degenerates", "Degeneracy", "GC-content (majority)", "Tm (majority)")) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), font_size = 11, full_width = TRUE)
}
```
<br>
**Table 3.** Forward primer, all variants.   
```{r}
knitr::kable(assay_report[[2]]$fwd$all_variants, digits = 2, align = "l", col.names = c("Variant", "GC-content", "Tm")) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), position = "left", font_size = 11, full_width = FALSE)
```
<br>
**Table 4.** Forward primer, mean values. 
```{r}
knitr::kable(assay_report[[2]]$fwd$mean_values, digits = 2, align = "l", col.names = "") %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), position = "left", font_size = 11, full_width = FALSE)
```
<br>
**Table 5.** Reverse primer, summary. 
```{r}
if (any(grepl("^pm_", names(assay_report[[1]][[1]])))) {
knitr::kable(assay_report[[1]]$reverse, digits = 2, col.names = c("Begin", "End", "Length", "Sequence (majority)", "Sequence (IUPAC)", "Degenerates", "Degeneracy", "GC-content (majority)", "Tm (majority)", "Perfect matches (majority)", "Perfect matches (IUPAC)")) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), font_size = 11, full_width = TRUE)
} else {
knitr::kable(assay_report[[1]]$reverse, digits = 2, col.names = c("Begin", "End", "Length", "Sequence (majority)", "Sequence (IUPAC)", "Degenerates", "Degeneracy", "GC-content (majority)", "Tm (majority)")) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), font_size = 11, full_width = TRUE)
}
```
<br>
**Table 6.** Reverse primer, all variants. 
```{r}
knitr::kable(assay_report[[2]]$rev$all_variants, digits = 2, col.names = c("Variant", "GC-content", "Tm")) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), position = "left", font_size = 11, full_width = FALSE)
```
<br>
**Table 7.** Reverse primer, mean values.
```{r}
knitr::kable(assay_report[[2]]$rev$mean_values, digits = 2, align = "l", col.names = "") %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), position = "left", font_size = 11, full_width = FALSE)
```
<br>
**Table 8.** Probe, summary. 
```{r}
if (any(grepl("probe", names(assay_report[[1]])))) {
if (any(grepl("^pm_", names(assay_report[[1]][[1]])))) {
knitr::kable(assay_report[[1]]$probe, digits = 2, col.names = c("Begin", "End", "Length", "Sequence (majority)", "Sequence (IUPAC)", "Degenerates", "Degeneracy", "GC-content (majority)", "Tm (majority)", "Sense", "Perfect matches (majority)", "Perfect matches (IUPAC)")) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), font_size = 11, full_width = TRUE)
} else { knitr::kable(assay_report[[1]]$probe, digits = 2, col.names = c("Begin", "End", "Length", "Sequence (majority)", "Sequence (IUPAC)", "Degenerates", "Degeneracy", "GC-content (majority)", "Tm (majority)")) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), font_size = 11, full_width = TRUE)

}
}
```
<br>
**Table 9.** Probe, all variants. 
```{r}
if (any(grepl("probe", names(assay_report[[1]])))) {
knitr::kable(assay_report[[2]]$pr$all_variants, digits = 2, align = "l", col.names = c("Variant", "GC-content", "Tm")) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), position = "left", font_size = 11, full_width = FALSE)
}
```
<br>
**Table 10.** Probe, mean values. 
```{r}
if (any(grepl("probe", names(assay_report[[1]])))) {
knitr::kable(assay_report[[2]]$pr$mean_values, digits = 2, col.names = "") %>%
  kableExtra::kable_styling(bootstrap_options = c("striped"), position = "left", font_size = 11, full_width = FALSE)
}
```
<br>

#### Match report 

<br>
**Table 11.** Perfectly matching sequences, majority oligos. 
```{r}
if (length(assay_report) == 3) {
  knitr::kable(assay_report[[3]]$matches_perfectly_majority, col.names = "") %>% kableExtra::kable_styling(bootstrap_options = c("striped"), position = "left", font_size = 11, full_width = FALSE)
}
```
<br>
**Table 12.** Mismatching sequences, majority oligos. TRUE = matching, FALSE = mismatching. 
```{r}
if (length(assay_report) == 3) {
knitr::kable(assay_report[[3]]$mismatches_majority) %>% kableExtra::kable_styling(bootstrap_options = c("striped"), position = "left", font_size = 11, full_width = FALSE)
}
```
<br>
**Table 13** Perfectly matching sequences, IUPAC oligos.
```{r}
if (length(assay_report) == 3) {
knitr::kable(assay_report[[3]]$matches_perfectly_iupac, col.names = "") %>% kableExtra::kable_styling(bootstrap_options = c("striped"), position = "left", font_size = 11, full_width = FALSE)
}
```
<br>
**Table 14.** Mismatching sequences, IUPAC oligos. TRUE = matching, FALSE = mismatching. 
```{r}
if (length(assay_report) == 3) {
knitr::kable(assay_report[[3]]$mismatches_iupac) %>% kableExtra::kable_styling(bootstrap_options = c("striped"), position = "left", font_size = 11, full_width = FALSE)
}
```
