---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

<!-- badges: start -->
[![R build status](https://github.com/sofpn/rprimer/workflows/R-CMD-check/badge.svg)](https://github.com/sofpn/rprimer/actions)
[![Codecov test coverage](https://codecov.io/gh/sofpn/rprimer/branch/master/graph/badge.svg)](https://codecov.io/gh/sofpn/rprimer?branch=master)
<!-- badges: end -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

## Installation

rprimer can be installed from [GitHub](https://github.com/) with:

  ``` r
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("sofpn/rprimer")
```
``` {r, message=FALSE, warn=FALSE, echo=FALSE}
devtools::load_all(".")
```

```{r, warn=FALSE, message=FALSE}
library(rprimer)
```

## Overview 

rprimer provides tools for designing degenerate DNA oligos for sequence variable targets.

The package contains five functions:

* `consensusProfile()`
* `oligos()` 
* `assays()`
* `checkMatch()`
* `plotData()`

These functions are interconnected in a workflow. The workflow can be run either directly from the R console or through a Shiny application. The application is loaded by typing `runRprimerApp()` in the R console. It can also be found online at: **LANK**

## Workflow 

### Import alignment

The first step is to import an alignment with target sequences of interest. This is done by using `readDNAMultipleAlignment()` from the Biostrings package.

The file "example_alignment.txt" contains an alignment of 200 hepatitis E virus sequences. 

```{r}
infile <- system.file("extdata", "example_alignment.txt", package = "rprimer")

myAlignment <- Biostrings::readDNAMultipleAlignment(infile, format = "fasta")
```

### Step 1: `consensusProfile` 

`consensusProfile()` takes a `Biostrings::DNAMultipleAlignment` as input and returns all the information needed for the subsequent design process.

```{r}
myConsensusProfile <- consensusProfile(myAlignment, ambiguityThreshold = 0.05)
```

Results (row 100-110):

```{r, echo=FALSE}
knitr::kable(myConsensusProfile[100:110, ], digits = 2)
```

The results can be visualized with `plotData()`. You can either plot the entire genome:

```{r, fig.width=12, fig.height=6}
plotData(myConsensusProfile)
```

Or zoom into a specific region of interest: 

```{r, fig.width=12, fig.height=6}
## Select position 5000 to 5500 
roi <- myConsensusProfile[myConsensusProfile$position >= 5000 & myConsensusProfile$position <= 5500, ]

plotData(roi)
```

The nucleotide distribution can be shown by specifying `type = "nucleotide`:

```{r, fig.width=12, fig.height=6}
## Select position 150 to 170
selection <- myConsensusProfile[myConsensusProfile$position >= 150 & myConsensusProfile$position <= 170, ]

plotData(selection, type = "nucleotide")
```

### Step 2: `oligos`

The next step is to design oligos. You can either use the default settings as below, or adjust them (see the package vignette or `?oligos` for more information). 

``` {r}
myOligos <- oligos(myConsensusProfile)
```

Results (first six rows): 

```{r, echo=FALSE}
knitr::kable(head(myOligos), digits = 2)
```

The results can be visualized as a dashboard, using `plotData()`: 

```{r, fig.width=12, fig.height=8}
plotData(myOligos)
```

### Step 3: `assays`

`assays()` finds pairs of forward and reverse primers and combines them with probes, if probes are present in the input dataset. You can either use the default settings as below, or adjust the design constraints (see the package vignette or `?assays` for more information).

```{r}
myAssays <- assays(myOligos)
```

Results (first six rows):

```{r, echo=FALSE}
knitr::kable(head(myAssays), digits = 2)
```

The assays can be visualized using `plotData()`: 

```{r, fig.width=12, fig.height=6}
plotData(myAssays)
```

### Check match 

`checkMatch()` shows the proportion and names of the target sequences in the input alignment that match with the generated oligos or assays. As an example, we check six randomly selected oligos from `myOligos`: 

```{r}
selection <- sample(seq_len(nrow(myOligos)), size = 6)

matchTableOligos <- checkMatch(myOligos[selection, ], target = myAlignment)
```

Results:

```{r, echo=FALSE}
knitr::kable(matchTableOligos, digits = 2)
```

The match table can be visualized using `plotData()`: 

```{r, fig.width=12, fig.height=6}
plotData(matchTableOligos)
```

## More information

The package vignette contains more information. It is loaded by `browseVignettes("rprimer")`.

## Citation

To cite this package, please use: `citation("rprimer")`.
