---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

  <!-- badges: start -->
  <!-- badges: end -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

## To do

* Oligos=describe design process 
* Classes=set validity
* All=test

## Installation

You can install rprimer from [GitHub](https://github.com/) with:

  ``` r
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("sofpn/rprimer")
```
``` {r, message=FALSE, warn=FALSE}
# library(rprimer)
devtools::load_all(".")
```

## Overview 

rprimer provides tools for designing broadly reactive primers, probes and (RT)-(q/d)PCR assays from a multiple DNA sequence alignment. 

The design process is built on three functions:

* `consensusProfile()`
* `oligos()` 
* `assays()`

## Workflow 

### Import alignment

The first step is to import an alignment with target sequences of interest and, if preferred, mask positions with e.g. high gap frequency. `readDNAMultipleAlignment()` and `maskGaps()` from Biostrings do the work for this part. The file "example_alignment.txt" contains an alignment of 200 hepatitis E virus sequences. 

```{r}
infile <- system.file("extdata", "example_alignment.txt", package = "rprimer")

myAlignment <- Biostrings::readDNAMultipleAlignment(infile, format = "fasta")
myAlignment <- Biostrings::maskGaps(myAlignment, 
                                    min.fraction = 0.5, 
                                    min.block.width = 1) 
```

### Step 1: `consensusProfile` 

`consensusProfile()` takes a `Biostrings::DNAMultipleAlignment` as input and returns all the information needed for the subsequent design process.

```{r}
myConsensusProfile <- consensusProfile(myAlignment, ambiguityThreshold = 0.05)
```

Results (first six rows):

```{r, echo=FALSE}
knitr::kable(head(myConsensusProfile), digits = 2)
```

The results can be visualized with `plotData()`. You can either plot the entire genome:

```{r, fig.width=12, fig.height=6}
plotData(myConsensusProfile)
```

Or zoom into a specific region of interest: 

```{r, fig.width=12, fig.height=6}
roi <- myConsensusProfile[myConsensusProfile$position >= 5000 & myConsensusProfile$position <= 5050, ]
plotData(roi)
```

The nucleotide distribution can be shown by specifying `type = "nucleotide`:

```{r, fig.width=12, fig.height=6}
plotData(roi, type = "nucleotide")
```

### Step 2: `oligos`

The next step is to design oligos. You can either use the default settings: 

``` {r}
myOligos <- oligos(myConsensusProfile)
```

Or adjust the constraints (see the package vignette or `?rprimer::oligos` for more information), for example: 

``` {r}
# myOligos2 <- oligos(myConsensusProfile, 
#                     maxDegeneracyPrimer = 16,
#                     tmRangePrimer = c(58, 60),
#                     conservedThreeEndPrimer =  FALSE,
#                     gcClampPrimer = FALSE,
#                     avoidThreeEndRunsPrimer = FALSE,
#                     probe = FALSE)
```

Results (first six rows): 

```{r, echo=FALSE}
knitr::kable(head(myOligos), digits = 2)
```

The results can be visualized using `plotData()`. 

```{r, fig.width=12, fig.height=8}
plotData(myOligos)
```

### Step 3: `assays`

`assays()` finds pairs of forward and reverse primers and combines them with probes, if probes are present in the dataset. You can either use default settings (as below), or customize the amplicon length and maximum allowed difference between the primers, and between the primer pair and probe. 

```{r}
myAssays <- assays(myOligos)
```

Results (first six rows):

```{r, echo=FALSE}
knitr::kable(head(myAssays), digits = 2)
```

The assays can be visualized using `plotData()`: 

```{r, fig.width=12, fig.height=6}
plotData(myAssays)
```

## More information

The package vignette contains more information. It is loaded by `browseVignettes("rprimer")`.
