---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

  <!-- badges: start -->
  [![R build status](https://github.com/sofpn/rprimer/workflows/R-CMD-check/badge.svg)](https://github.com/sofpn/rprimer/actions)
  <!-- badges: end -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

**to do: validate objs, kolla documentation/alla defaults, write tests, skriv vingette, installera, jmfr med existerande (primer3, open primer), skicka**

### Installation

You can install rprimer from [GitHub](https://github.com/) with:

  ``` r
# install.packages("devtools")
devtools::install_github("sofpn/rprimer")
```
Initial setup for the code in this document: 

``` {r, message=FALSE, warn=FALSE}
# library(rprimer)
devtools::load_all(".")
library(magrittr) # Required for the pipe operator 
library(Biostrings) # Required to import alignment
```

### Overview 

rprimer provides tools for designing (RT)-(q/dd)PCR assays from multiple DNA sequence alignments. The design process is built on three functions: 

* `getConsensusProfile()`: returns an `RprimerProfile` object, which is used as input for;
* `getOligos()`: returns an `RprimerOligo` object, which is used as input for; 
* `getAssays()`: returns an `RprimerAssay` object.

The `Rprimer`-classes are extensions of the `DataFrame` class from S4Vectors, and behave in a similar manner as traditional data frames (with methods for e.g. [, head, tail), and can be coerced to such by using `as.data.frame()`.

### Import data 

The first step is to import an alignment with target sequences of interest and, if preferred, mask positions with high gap frequency (caution when removing gaps though!). `readDNAMultipleAlignment()` and `maskGaps()` from Biostrings do the work for this part. 

Here, I  want to design an RT-(q/d)PCR assay for detection of hepatitis E virus, which is a highly variable RNA virus. The file "example_alignment.txt" is provided and contains 100 hepatitis E virus sequences. 

```{r}
infile <- system.file('extdata', 'example_alignment.txt', package = 'rprimer')

myAlignment <- infile %>%
  Biostrings::readDNAMultipleAlignment(., format = "fasta") %>%
  Biostrings::maskGaps(., min.fraction = 0.5, min.block.width = 1) 
```

### Step 1: `getConsensusProfile` 

`getConsensusProfile()` takes a `Biostrings::DNAMultipleAlignment` object as input and returns all the information needed for the subsequent design process. Masked positions are excluded. 

```{r}
myConsensusProfile <- getConsensusProfile(myAlignment, iupacThreshold = 0.05)
```

Output: 

```{r, echo=FALSE}
knitr::kable(head(myConsensusProfile), digits = 2)
```

Some comments on the data: 

* Majority refers to the majority consensus sequence, which is the most frequently occurring base. Identity is the proportion of that base when gaps and other bases than A, C, G and T are not taken into account. 

* The IUPAC consensus sequence includes wobble bases according to the IUPAC-nomenclature. It includes all DNA bases (A, C, G, T) that occurs with a frequency higher than the stated `iupacThreshold`. **The iupac treshold has implications for downstream oligo design.**

* Entropy refers to Shannon entropy, which is a measurement of variability. A value of zero indicate no variability and a high value indicate high variability. Gaps and other bases than A, C, G and T are not included in this calculation. 

The target genome can be visualized with `plotData()`, and specific regions can be highlighted using the optional arguments `shadeFrom` and `shadeTo`. We can see that the most conserved region of the hepatitis E virus genome is between position 5000-5500:

```{r, fig.width=12, fig.height=6}
plotData(myConsensusProfile, shadeFrom = 5000, shadeTo = 5500)
```

The black lines represent centered running averages and the blue dots represent the value at each position. 

You can also subset the data to look at a certain region in more detail: 

```{r, fig.width=12, fig.height=6}
mySubset <- myConsensusProfile[myConsensusProfile$position > 5000 & myConsensusProfile$position < 5500, ]

plotData(mySubset)
```

### Step 2: `getOligos`

`getOligos()` searches for oligos from an `RprimerProfile`-object. All oligos are shown in both majority and IUPAC format. The oligos are designed from the following constraints:  

* `lengthPrimer` Primer length, defaults to `18:22`.
* `maxGapFrequencyPrimer` Maximum gap frequency for primers, defaults to `0.1`. 
* `maxDegeneracyPrimer` Maximum number of degenerate variants of each primer, defaults to `4`. 
* `gcClampPrimer` If primers must have a GC-clamp, defaults to `TRUE`. 
* `avoid3endRunsPrimer` If primers with more than two runs of the same nucleotide at the terminal 3' end should be avoided, defaults to `TRUE`.
* `minEndIdentityPrimer` Optional. Minimum allowed identity at the 3' end (i.e. the last five bases). E.g., if set to `1`, only primers with complete target conservation at the 3' end will be considered.
* `gcRangePrimer` GC-content-range for primers, defaults to `c(0.45, 0.55)`.
* `tmRangePrimer` Melting temperature (Tm) range for primers, defaults to `c(55, 65)`. Tm is calculated using the nearest-neighbor method. See `?rprimer::getOligos` for a detailed description and references.
* `concPrimer` Primer concentration in nM (for Tm calculation), defaults to `500`.
* `probe` If probes should be designed as well, defaults to `TRUE`.
* `lengthProbe` Probe length, defaults to `18:22`.
* `maxGapFrequencyProbe` Maximum gap frequency for probes, defaults to `0.1`. 
* `maxDegeneracyProbe` Maximum number of degenerate variants of each probe, defaults to `4`. 
* `avoid5endGProbe` If probes with a G at the terminal 5' end should be avoided, defaults to `TRUE`.
* `gcRangeProbe` GC-content-range for probes, defaults to `c(0.45, 0.55)`.
* `tmRangeProbe` Melting temperature (Tm) range for probes, defaults to `c(55, 70)`. 
* `concProbe` Probe concentration in nM (for Tm calculation), defaults to `250`.
* `concNa` Sodium ion concentration in the PCR reaction (for Tm calculation), defaults to `0.05` M (50 mM).
* `showAllVariants` If sequence, GC-content and Tm should be presented for all variants of each oligo (in case of degenerate bases).`TRUE` (slower) or `FALSE` (faster), defaults to `TRUE`. 

In addition, `getOligos()` avoids:

* Majority oligos with more than than three consecutive runs of the same dinucleotide (e.g. "TATATATA")
* Majority oligos with more than four consecutive runs of the same nucleotide  (e.g. "AAAAA")
* Majority oligos that are duplicated (to prevent binding at several places on the genome)

A warning message will return if no oligos are found. 

Below, I design both primers and probes.

``` {r}
myOligos <- getOligos(myConsensusProfile,
                      lengthPrimer = 18:22,
                      maxGapFrequencyPrimer = 0.1,
                      maxDegeneracyPrimer = 4,
                      gcClampPrimer = TRUE,
                      avoid3EndRunsPrimer = TRUE,
                      minEndIdentityPrimer = 0.98,
                      gcRangePrimer = c(0.45, 0.65),
                      tmRangePrimer = c(55, 65),
                      concPrimer = 500,
                      probe = TRUE,
                      lengthProbe = 16:24,
                      maxGapFrequencyProbe = 0.1,
                      maxDegeneracyProbe = 4,
                      avoid5EndGProbe = TRUE, 
                      gcRangeProbe = c(0.45, 0.65),
                      tmRangeProbe = c(55, 70),
                      concProbe = 250,
                      concNa = 0.05,
                      showAllVariants = TRUE)
```

Output: 

```{r, echo=FALSE}
knitr::kable(myOligos[10:15, ], digits = 2)
```

NA means that the oligo was invalid. 

For the `RprimerProfile` object, it is possible to inspect the nucleotide distribution with `plotNucleotides()`. For instance, we can select to plot the region where the first primer appears. `rc` regulates whether the sequence should be displayed as a reverse complement or not.

```{r, fig.align="center", fig.width=5, fig.height=4}
## Select position 40-61 in the genome 
myOligoRegion <- myConsensusProfile[myConsensusProfile$position >= 40 & myConsensusProfile$position <= 61, ] 

plotNucleotides(myOligoRegion, rc = FALSE) 
```
  
All oligo candidates can be visualized using `plotData()`:

```{r, fig.width=12, fig.height=8}
plotData(myOligos)
```

It is also possible, by subsetting, to select oligos targeting a specific region: 

```{r, fig.width=12, fig.height=8}
plotData(myOligos[myOligos$start > 5000 & myOligos$end < 6000, ])
```

### Step 3: `getAssays`

`getAssays()` finds pairs of forward and reverse primers and combines them with probes (if selected), based on the following constraints: 

* `length` Amplicon length, defaults to `65:120`.
* `maxTmDifferencePrimers` Maximum Tm difference between the two primers (absolute value, calculated for majority primers), defaults to `2`.
* `tmDifferencePrimersProbe` Acceptable Tm difference between the primers (average Tm of the primer pair) and probe, defaults to `c(0, 20)`. The Tm-difference is calculated by subtracting the Tm of the (majority) probe with the average Tm of the (majority) primer pair. A negative Tm-difference means that the Tm of the probe is lower than the average Tm of the primer pair.

An error message will return if no assays are found.

```{r}
myAssays <- getAssays(myOligos, 
                      length = 65:120,
                      maxTmDifferencePrimers = 2,
                      tmDifferencePrimersProbe = c(-2, 10))
```

Output: 

```{r}
knitr::kable(myAssays[10:15, ], digits = 2)
```

The output can be visualized using `plotData()`: 

```{r, fig.width=12, fig.height=6}
plotData(myAssays)
```

### Example data 

Example data sets of each class can be loaded by: 

```{r}
data("exampleRprimerProfile")
data("exampleRprimerOligo")
data("exampleRprimerAssay")
```

### Session info 

```{r}
sessionInfo()
```
