---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

  <!-- badges: start -->
  [![R build status](https://github.com/sofpn/rprimer/workflows/R-CMD-check/badge.svg)](https://github.com/sofpn/rprimer/actions)
  <!-- badges: end -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

### Installation

You can install the development version of rprimer 
from [GitHub](https://github.com/) with:

  ``` r
# install.packages("devtools")
devtools::install_github("sofpn/rprimer")
```
Initial setup for this session: 

``` {r, message=FALSE, warn=FALSE}
# library(rprimer)
devtools::load_all(".")
library(magrittr) ## Required for the pipe operator in this example
```

### Introduction 

The purpose of rprimer is to design (RT)-(q/dd)PCR assays from multiple 
sequence alignments. In this document, I demonstrate how to use the package 
by designing an RT-qPCR assay for detection of hepatitis E virus. 

### To start: Import alignment and mask positions with high gap frequency

For this part, I use previously existing functionality from the Biostrings-package. 

Import an alignment with target sequences of interest using 
`Biostrings::readDNAMultipleAlignment()`, and mask positions with high gap frequency
using `Biostrings::maskGaps()`.

The file "example_alignment.txt" is provided with the package and consists of 100
hepatitis E virus sequences. 

```{r}
infile <- system.file('extdata', 'example_alignment.txt', package = 'rprimer')

myAlignment <- infile %>%
  Biostrings::readDNAMultipleAlignment(., format = "fasta") %>%
  Biostrings::maskGaps(., min.fraction = 0.5, min.block.width = 1)
```

The alignment can be visualised with the `rpPlot()` function from this package.

### Step 1: `getAlignmentProfile()` 

`getAlignmentProfile()` is a wrapper around `Biostrings::consensusMatrix()`. 
It takes a `Biostrings::DNAMultipleAlignment` object as input and returns 
an `RprimerProfile` object, which contains a numeric matrix that 
holds the proportion of each nucleotide at each position in the alignment. 

```{r}
myAlignmentProfile <- getAlignmentProfile(myAlignment)
(myAlignmentProfile[ , 1:30]) ## View the first 30 bases 
```

The object can be visualized using `rpPlot()`. 
The `rc` option regulates whether it
should be displayed as a reverse complement or not. 

```{r, fig.align="center", fig.width=6, fig.height=6}
rpPlot(myAlignmentProfile[, 1:30], rc = FALSE) ## Plot the first 30 bases 
```

### Step 2: `getAlignmentProperties()`

`getAlignmentProperties()` takes an `RprimerProfile`-object as input and 
returns an `RprimerProperties`-object with information about majority and IUPAC
consensus base at each position, together with gap frequency, 
nucleotide identity and Shannon entropy (a measurement of variability). All 
nucleotides that occurs with a frequency higher than the `iupacThreshold` 
will be included in the IUPAC consensus base. 

```{r}
myAlignmentProperties <- getAlignmentProperties(
  myAlignmentProfile, iupacThreshold = 0.05 
)
head(myAlignmentProperties)
```
The object can be visualised with `rpPlot()`. 

### Step 3: `getOligos()`

`getOligos()` takes an `RprimerProperties` object as input and searches 
for oligos based on the following constraints: 

* `maxGapFrequency` Maximum gap frequency, defaults to 0.1 
* `length` Oligo length, defaults to `18:22`.
* `maxDegeneracy` Maximum number of degenerate variants of each oligo, 
defaults to `4`. 
* `gcClamp` If oligos must have a GC-clamp to be considered as valid
(recommended for primers), defaults to `TRUE`. 
* `avoid3endRuns` If oligos with more than two runs of the same nucleotide at 
the terminal 3' end should be excluded (recommended for primers), 
defaults to `TRUE`.
* `avoid5endG` If oligos with a G at the terminal 5' end should be avoided 
(recommended for probes), defaults to `FALSE`.
* `minEndIdentity` Optional. Minimum allowed identity at the 3' end 
(i.e. the last five bases). E.g., if set to `1`, only oligos with complete 
target conservation at the 3' end will be considered.
* `gcRange` GC-content-range, defaults to `c(0.45, 0.55)`.
* `tmRange` Melting temperature (Tm) range, defaults to `c(50, 65)`. 
Tm is calculated using the nearest-neighbor method. See `?rprimer::getOligos` 
for a detailed description and references.
* `concOligo` Oligo concentration (for Tm calculation), 
defaults to `5e-07` M (500 nM)
* `concNa` Sodium ion concentration (for Tm calculation), 
defaults to `0.05` M (50 mM).
* `showAllVariants` If sequence, GC-content and Tm should be presented for 
all variants of each oligo (in case of degenerate bases).`TRUE`
(slower) or `FALSE` (faster), defaults to `TRUE`. 

In addition, `get_oligos()` avoids:

* Oligos with more than than three consecutive runs of the same dinucleotide 
(e.g. "TATATATA")
* Oligos with more than four consecutive runs of the same nucleotide 
(e.g. "AAAAA")
* Oligos that are duplicated (to prevent binding at several places on the genome)

Below, I want to design both primers and probes. I use somewhat different 
settings for the two oligo types.

An error message will return if no oligos are found. 

``` {r}
myPrimers <- getOligos(myAlignmentProperties,
                       length = 18:22,
                       maxGapFrequency = 0.05,
                       maxDegeneracy = 4,
                       gcClamp = TRUE,
                       avoid3EndRuns = TRUE,
                       avoid5EndG = FALSE,
                       minEndIdentity = 0.99,
                       gcRange = c(0.40, 0.60),
                       tmRange = c(50, 65),
                       showAllVariants = TRUE)

myProbes <-  getOligos(myAlignmentProperties,
                       length = 16:22,
                       maxGapFrequency = 0.05,
                       maxDegeneracy = 4,
                       gcClamp = FALSE,
                       avoid3EndRuns = FALSE,
                       avoid5EndG = TRUE,
                       minEndIdentity = NULL,
                       gcRange = c(0.40, 0.60),
                       tmRange = c(50, 75),
                       showAllVariants = TRUE)
```

### Step 4: `getAssays()`
`getAssays()` finds pairs of forward and reverse primers and combines them with
probes (if selected). It takes `RprimerOligo` objects as input and returns an
`RprimerAssay` object. 

Assays are designed from the following constraints: 

  * `length` Amplicon length, defaults to `65:120`.
  * `maxTmDifferencePrimers` Maximum Tm difference between the two primers 
  (absolute value), defaults to `2`.
  * `tmDifferenceProbes` Acceptable Tm difference between the primers 
  (average Tm of the primer pair) and probe, defaults to `c(0, 20)`.

Candidate assays are displayed in a tibble. 
An error message will return if no assays are found.

```{r}
myAssays <- getAssays(primers = myPrimers, 
                      probes = myProbes)
```


That's it!  

For more detailed information, please see the help files or the package vingette.

### Session info 

```{r}
sessionInfo()
```

