---
title: "Design report"
output: html_document
params: 
  assay_selection: assay_selection
  sequence_profile: sequence_profile 
  sequence_properties: sequence_properties
  comment: comment  
   
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, eval = TRUE)
```
<br><br>
**Alignment size:** `r nrow(assay_selection$match_matrix[[1]])` sequences
<br>
**Length of alignment/region of interest:** `r ncol(sequence_profile)` nucleotides
<br>
**Report generated:** `r Sys.time()` 
<br>
**Comment:** `r params$comment`
<br><br>

### Assay information 

```{r}
assay_report <- print_assay_report(params$assay_selection)
```

```{r, fig.width=12, fig.height=8}
plot_assay_overview(params$sequence_properties, params$assay_selection)
```
<br><br>
**Fig. 1.** Assay location (blue) and target information. *Identity:* Proportion of the most frequently occuring base. *Entropy:* Shannon entropy, a measurement of variability. A value of zero indicate no variability and a high value indicate high variability. *GC:* GC-content. *Gaps:* Proportion of gaps. Vertical bars represent the value at each position. Black bars in the two uppermost figures represent values of 1 and 0, respectively (i.e. positions with complete conservation). Horizontal lines represent centered running averages.   
<br><br>
```{r, fig.width=12, fig.height=6}
plot_assay_details(params$sequence_profile, params$assay_selection)
```
<br><br>
**Fig. 2.** Proportion of each nucleotide in the alignment at the oligo binding sites. Sequences are displayed from 5'- to 3'-end. 
<br><br>
**Table 1**. Assay information. *Begin:* Where the assay region starts (with regards to the alignment/region of interest), *End:* Where the assay region ends. *Tm difference, primers:* Tm difference between the two primers (absolute value, in C). *Tm difference, probe-primers:* Tm of the probe subtracted by the average Tm of the primer pair. Note that tm-differences are calculated from majority oligos and may therefore be misleading for degenerate (IUPAC) oligos. *Degeneracy:* Total number of oligo variants in the assay. *Perfect match:* Proportion of perfectly matching sequences. 
<br>
```{r}
# Customise table options 
custom_table <- function(x, col.names, full_width = TRUE) {
  knitr::kable(x, "html", digits = 2, col.names = col.names, align = "l") %>%
    kableExtra::kable_styling(
      bootstrap_options = "basic", position = "left", 
      font_size = 11, full_width = full_width
    )
}

has_probe <- any(grepl("probe", names(assay_report[[1]])))

if (has_probe) {
  custom_table(
  assay_report[[1]]$general, 
  col.names = c(
    "Begin", "End", "Amplicon length", "Tm difference, primers", 
    "Tm difference, probe-primers", "Degeneracy", "Perfect match (majority)", 
    "Perfect match (IUPAC)"
  )
)
} else {
  custom_table(
  assay_report[[1]]$general, 
  col.names = c(
    "Begin", "End", "Amplicon length", "Tm difference, primers", 
     "Degeneracy", "Perfect match (majority)", 
    "Perfect match (IUPAC)"
  )
)
}

```
<br>
**Table 2.** Forward primer, summary.
<br>
```{r}
custom_table(
  assay_report[[1]]$forward, 
  col.names = c(
    "Begin", "End", "Length", "Sequence (majority)", "Sequence (IUPAC)", 
    "Degenerates", "Degeneracy", "GC-content (majority)", "Tm (majority)", 
    "Perfect matches (majority)", "Perfect matches (IUPAC)")
)
```
<br>
**Table 3.** Forward primer, all variants.   
<br>
```{r}
custom_table(
  assay_report[[2]]$fwd$all_variants,
  col.names = c("Variant", "GC-content", "Tm"), full_width = FALSE
)
```
<br>
**Table 4.** Forward primer, mean values. 
<br>
```{r}
custom_table(
  assay_report[[2]]$fwd$mean_values, col.names = "", full_width = FALSE
)
```
<br>
**Table 5.** Reverse primer, summary.
<br>
```{r}
custom_table(
  assay_report[[1]]$reverse, 
  col.names = c(
    "Begin", "End", "Length", "Sequence (majority)", "Sequence (IUPAC)", 
    "Degenerates", "Degeneracy", "GC-content (majority)", "Tm (majority)", 
    "Perfect matches (majority)", "Perfect matches (IUPAC)")
)
```
<br>
**Table 6.** Reverse primer, all variants. 
<br>
```{r}
custom_table(
  assay_report[[2]]$rev$all_variants,
  col.names = c("Variant", "GC-content", "Tm"), full_width = FALSE
)

```
<br>
**Table 7.** Reverse primer, mean values.
<br>
```{r}
custom_table(
  assay_report[[2]]$rev$mean_values, col.names = "", full_width = FALSE
)
```
<br>
**Table 8.** Probe, summary. 
<br>
```{r}
if (has_probe) {
custom_table(
  assay_report[[1]]$probe, 
  col.names = c(
    "Begin", "End", "Length", "Sequence (majority)", 
    "Sequence (IUPAC)", "Degenerates", "Degeneracy", "GC-content (majority)", 
    "Tm (majority)", "Perfect matches (majority)", 
    "Perfect matches (IUPAC)", "Sense"
  )
)
}
```
<br>
**Table 9.** Probe, all variants. 
<br>
```{r}
if (has_probe) {
  custom_table(
  assay_report[[2]]$pr$all_variants,
  col.names = c("Variant", "GC-content", "Tm"), full_width = FALSE
)
}
```
<br>
**Table 10.** Probe, mean values. 
<br>
```{r}
if (has_probe) {
 custom_table(
   assay_report[[2]]$rev$mean_values, col.names = "", full_width = FALSE
 )
}
```
<br>

### Amplicon information 

```{r}
# Get the consensus amplicon sequence 
consensus <- paste(sequence_properties$majority[assay_report[[1]]$general$begin:assay_report[[1]]$general$end], collapse = "")
```

**Sequence:** `r consensus`
<br>
**Length:** `r nchar(consensus)`
<br>
**GC-content:** `r round(gc_content(consensus), 2)*100` %
<br>
**Does the amplicon contain more than 3 consecutive G:s?** `r find_g_repeats(consensus)` 
<br><br>
The data is based on the majority consensus sequence. 

### Match information (in silico inclusivity)

```{r, fig.width=12, fig.height=12}
plot_match_matrix(assay_selection)
```
<br><br>
**Fig. 3.** Matching and mismatching target sequences.
<br>

### Target sequences 

The assay was designed from the following target sequences: 
<br><br>
`r rownames(assay_selection$match_matrix[[1]])`.
