---
title: "rprimer"
author: "Sofia"
package: rprimer
output: BiocStyle::html_document
bibliography: bibliography.bibtex
vignette: >
  %\VignetteIndexEntry{Getting started}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

Reverse transcription (RT) PCR based methods are important cornerstones of both clinical diagnosis and environmental surveillance of viruses. Successful RT-PCR analysis depends on several factors, but oligo (primer and probe) design is arguably the most critical part of assay development [@Bustin]. Poorly designed primers may reduce the efficiency of the RT or PCR and poorly designed probes may not hybridize properly with the target, which both can lead to underestimation of the target concentration or false negative results.

The ideal oligo binds exclusively to a conserved part of the target of interest and does not fold or form stable interactions (primer dimers) with other oligos. However, it is often both difficult and time consuming to identify target regions and oligo combinations that meets all these criteria. Moreover, many viruses, especially RNA viruses, evolve quickly and thereby pose significant sequence variation between strains. This makes sequence conservation analysis a key step of the oligo design process, since primers or probes that captures only a part of the target variation can cause severe detection and quantification bias.

A common approach for amplifying sequence variable regions is to use degenerate oligos. A degenerate oligo has one or more positions with several possible bases, and the degeneracy refers to the number of unique sequence variants it encompasses. The ambiguous bases are specified by a code system determined by IUPAC. For instance, the IUPAC consensus base K can be either G or T, and N can be A, C, G or T. For some applications, it is not uncommon to use primers with up to several hundreds or thousands of sequence variants. However, as the degeneracy increases, the fraction of oligos that match perfectly to the target sequence and the synthesized products decreases, which risks to reduce the amplification efficiency [@Rose]. A preferred design strategy would therefore be to find oligo binding regions with as low degeneracy as possible.

rprimer provides tools for analyzing sequence conservation and designing degenerate primers and probes for sequence variable viruses. A multiple DNA sequence alignment with target sequences is used as input and the outputs consist of data frames (tabular data) and dashboard-like plots. In this document, I describe and demonstrate how to use the package by designing broadly-reactive RT-(q/d)PCR oligos and assays for detection of hepatitis E virus (HEV), which is a sequence variable RNA-virus and a common foodborne pathogen.  

# Installation

rprimer can be installed from [GitHub](https://github.com/) with:

``` r
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("sofpn/rprimer")
```
```{r, internal setup, echo=FALSE, warning=FALSE, message=FALSE}
library(kableExtra)
#devtools::load_all(".")
```
```{r setup, message=FALSE, warning=FALSE}
library(rprimer)
```

# Overview

The package contains five functions:

* `consensusProfile()`
* `oligos()`  
* `assays()` 
* `plotData()`
* `convertToDNAStringSet()`

# Workflow

## Collection of target sequences and multiple alignment 

The first step is to identify and align target sequences of interest. The file “example_alignment.txt” contains an alignment of 200 HEV sequences, and the filepath can be retrieved by:  

```r
system.file("extdata", "example_alignment.txt", package = "rprimer")
```

These sequences were collected by searching for “[a-z]”[porgn:__txid1678143] (the taxid of Ortohepevirus A, i.e., HEV) on the nucleotide database of NCBI GenBank, and filtering for sequence lengths > 6000, to retrieve complete and nearly complete genomes. The search had 674 hits (as of January 19, 2021). The sequences were downloaded (as complete record, fasta-file) and run through Hepatitis E virus genotyping tool version 0.1 (https://www.rivm.nl/mpf/typingtool/hev/, accessed January 19, 2021) to verify that they were correctly classified as Ortohepevirus A. Sequences that were not classified as Ortohepevirus A were removed, and the remaining ones (n = 642) were reduced to 200 by random selection, to not provide a too large example dataset for the package. The sequences were then aligned using the online version of mafft version 7 (https://mafft.cbrc.jp/alignment/server/, accessed January 19, 2021), using Auto settings.

## Data import 

Before the design process can begin, the alignment must be imported to R. This is done using the `readDNAMultipleAlignment()` function from Biostrings [@PagesBios]. The input file can contain from one up to several thousands of sequences.

```{r}
filepath <- system.file("extdata", "example_alignment.txt", package = "rprimer")

myAlignment <- Biostrings::readDNAMultipleAlignment(filepath, format = "fasta")
```

## Design

### Step 1: `consensusProfile` 

Once the alignment is imported, the first step is to generate a consensus profile, with the following information: 

* The proportion of the bases A, C, G, T and “other”
* The proportion of gaps (-)
* Majority consensus base, which is the most frequently occurring DNA base
* Identity, which is the proportion of the majority consensus base when gaps and ambiguous bases ("other") in the target alignment are excluded
* IUPAC consensus character (ambiguous base). The user can specify an “ambiguity threshold” from 0 to 0.2, and all bases with a relative frequency higher than that will be included in the consensus character. A threshold of 0 (default) will catch all the variation, but with the potential cost of generating oligos with high degeneracy. A higher threshold will catch most of the variation, but with the potential cost of missing minor sequence variations (this information is provided by the coverage, see below)
* Shannon entropy, which is a measurement of variability [@shannon1951prediction]. A value of zero indicate complete conservation, and a high value indicate high variability. Gaps and ambiguous bases ("other") in the target alignment are not included in this calculation
* Coverage, which describes the total proportion of bases that are "covered" by the IUPAC consensus character. Gaps and ambiguous bases ("other") in the target alignment are not included in this calculation

The consensus profile is generated as follows:

```{r}
myConsensusProfile <- consensusProfile(myAlignment, ambiguityThreshold = 0.05)
```

Results (row 100-110):

```{r, echo=FALSE}
myConsensusProfile[100:110, ] %>%
  kbl(., digits = 2) %>%
  kable_material(c("striped", "hover"), position = "right") %>%
  scroll_box(width = "650px", height = "300px")
```

The results can be visualized with `plotData()`. 

```{r, fig.width=12, fig.height=6}
plotData(myConsensusProfile)
```

The dots represent the value at each position and the black lines represent centered running averages. High identity and low entropy values indicate high sequence conservation.

It is possible to zoom into a specific region of interest:

```{r, fig.width=12, fig.height=6}
## Select position 1000 to 1500 in the consensus profile 
selection <- myConsensusProfile[
  myConsensusProfile$position >= 1000 & myConsensusProfile$position <= 1500, 
  ]

plotData(selection)
```

And to visualize the nucleotide distribution, by specifying `type = nucleotide`: 

```{r, fig.width=12, fig.height=6}
## Select position 1000 to 1020 in the consensus profile 
ntSelection <- myConsensusProfile[
  myConsensusProfile$position >= 1000 & myConsensusProfile$position <= 1020, 
  ]

plotData(ntSelection, type = "nucleotide")
```

### Step 2: `oligos`

The next step is to design oligos. Primers must be designed and probes are optional. Primers can be generated from one of the two following strategies: 

* The **ambiguous strategy** (default) generates primers from the IUPAC consensus sequence, which means that degenerate bases can occur at any position in the primer

* The **mixed strategy** is inspired by the Consensus-Degenerate Hybrid Oligonucleotide Primer (CODEHOP) principle [@Rose2]. Here, primers are generated from both the majority and the IUPAC consensus sequence, and consist of a shorter degenerate part at the 3' end (~1/3 of the primer, targeting a conserved region), and a longer consensus part at the 5' end (~2/3 of the primer), which instead of having ambiguous bases contains the most probable nucleotide at each position. The idea is that the degenerate 3' end part will bind specifically to the target sequence in the initial PCR cycles, and promote amplification in spite of eventual mismatches at the 5' consensus part (since 5' end mismatches are generally less detrimental than 3' end mismatches) [@Rose2]. The generated products will match the 5' ends of all primers perfectly, which allows for efficient amplification of the PCR products in the later cycles. To provide a sufficiently high tm in spite of mismatches, it is recommended to design relatively long primers (> 25 bases) when using this strategy  

Probes are always designed using the ambiguous strategy. 

The design settings (i.e. the arguments of `oligos()`) are listed below:

* `x`: the consensus profile from Step 1
* `maxGapFrequency`: maximum gap frequency (in the alignment) for primers and probes, defaults to `0.05`
* `lengthPrimer`: primer length, defaults to `18:22`
* `maxDegeneracyPrimer`: maximum degeneracy for primers, defaults to `4`
* `gcClampPrimer`: if primers should have a GC clamp, defaults to `TRUE`. A GC clamp is here defined as the presence of two to three G or C:s within the last five bases (3' end) of the oligo. A GC clamp helps to promote specific binding of the 3' end
* `avoidThreeEndRunsPrimer`: if primers with more than two runs of the same nucleotide at the terminal 3' end should be avoided (to reduce the risk of mispriming), defaults to `TRUE`
* `minThreeEndCoveragePrimer`: minimum allowed coverage at the 3' end (the last five bases). If set to `1`, all bases of the 3' end must cover all sequence variants in the target alignment. Defaults to `0.98` 
* `gcRangePrimer`: GC-content range of primers (proportion), defaults to `c(0.40, 0.65)`
* `tmRangePrimer`: melting temperature range for primers, defaults to `c(55, 65)`. Melting temperature is calculated for perfectly matching oligo-target duplexes using the nearest neighbor method [@SantaLuciaUnified], using formula, salt correction method and table values as described in [@santalucia2004thermodynamics]
* `concPrimer`: primer concentration in nM (for tm calculation), defaults to `500`
* `designStrategyPrimer`: design strategy for primers, `"ambiguous"` or `"mixed"`, defaults to `"ambiguous"` 
* `probe`: if probes should be designed as well, defaults to `TRUE`
* `lengthProbe`: defaults to `18:22`
* `maxDegeneracyProbe`: defaults to `4`
* `avoidFiveEndGProbe`: if hydrolysis probes with a G at the terminal 5' end should be avoided (to prevent quenching of the 5' flourophore), defaults to `TRUE` 
* `gcRangeProbe`: defaults to `c(0.40, 0.65)`
* `tmRangeProbe`: defaults to `c(55, 70)`
* `concProbe`: defaults to `250`
* `concNa`: sodium ion concentration (in M) in the PCR reaction (for calculation of tm and delta G), defaults to `0.05` (50 mM)
* `temperature`: annealing temperature (in C), for calculation of delta G. Delta G is calculated for perfectly matching oligo-target duplexes using the nearest neighbor method [@SantaLuciaUnified], using formula, salt correction method and table values as described in [@santalucia2004thermodynamics]

All sequence variants of an oligo must fulfill all the specified design constraints to be considered. 

Oligos with sequence variants containing more than four consecutive runs of the same nucleotide (e.g. "AAAAA") and/or more than three consecutive runs of the same di-nucleotide (e.g. "TATATATA") are excluded from consideration.

An error message will return if no oligos are found. If so, a good idea could be to re-run the process from Step 1 and increase the `ambiguityThreshold` in `consensusProfile()`, and/or relax the design constraints above. 

**Design with default settings**

In the first example, I show how to design oligos with the default settings as described above.

``` {r}
myOligos <- oligos(myConsensusProfile)
```

Results (first ten rows):

```{r, echo=FALSE}
myOligos[1:10, ] %>%
  kbl(., digits = 2) %>%
  kable_material(c("striped", "hover"), position = "right") %>%
  scroll_box(width = "650px", height = "300px")
```

The man page (`?oligos`) contains a detailed description of each variable, but what may not be entirely self-explanatory is that some variables (`sequence` `gcContent`, `tm` and `deltaG`) can hold several values in each row. All values within a specific row can be retrieved by:   

```{r}
myOligos$sequence[[1]] ## All sequence variants of the first oligo (i.e., first row) 
myOligos$gcContent[[1]] ## GC-content of all variants of the first oligo 
myOligos$tm[[1]] ## Tm of all variants of the first oligo 
myOligos$deltaG[[1]] ## Delta G of all variants of the first oligo (in kcal/mole)
```

The primer and probe candidates can be visualized using `plotData()`:

```{r, fig.align="center", fig.width=12, fig.height=8}
plotData(myOligos)
```

Moreover, each oligo is scored upon on their average identity, coverage, degeneracy, GC-content and tm range (see the 'score' variable in the table above). The value can range from 0 to 12, where 0 is best. Detailed information about the scoring system can be found at the man page (`?oligos`).

As an example, I select to look at the best scoring oligo(s):

```{r, fig.align="center", fig.width=12, fig.height=8}
## Get the minimum score from myOligos 
bestOligoScore <- min(myOligos$score)
bestOligoScore

## Make a subset that only oligos with the best score are included 
oligoSelection <- myOligos[myOligos$score == bestOligoScore, ]

## Plot these oligos 
plotData(oligoSelection)
```

It is also possible to select a specific oligo and plot the nucleotide distribution of the binding region, by subsetting the consensus profile from Step 1. 

```{r}
## Get the binding region of the first oligo in the selection above (first row): 
bindingRegion <- myConsensusProfile[
  myConsensusProfile$position >= oligoSelection[1, ]$start & 
    myConsensusProfile$position <= oligoSelection[1, ]$end,
  ]
```

It can be plotted in forward direction: 

```{r}
plotData(bindingRegion, type = "nucleotide")
```

Or as reverse complement, by specifying `rc = TRUE`: 

```{r}
plotData(bindingRegion, type = "nucleotide", rc = TRUE)
```

**Design with modified settings**

In the next example, I show how to modify some of the design constraints: 

```{r}
myOligosModified <- oligos(myConsensusProfile, 
                           maxDegeneracyPrimer = 32, 
                           tmRangePrimer = c(50, 70),
                           lengthPrimer = 16:24, 
                           maxDegeneracyProbe = 1)
```

Now, I get more primer candidates (due to relaxed constraints) and less probe candidates (due to tighter constraints):

```{r, fig.align="center", fig.width=12, fig.height=8}
plotData(myOligosModified)
```

**Select a region of interest**

Below, I show how to select a particular region of interest (position 1000 to 4000):

```{r}
roi <- myConsensusProfile[
  myConsensusProfile$position >= 1000 & myConsensusProfile$position <= 4000, 
  ]
```

I now design mixed primers for this region, and select to not design probes:

```{r, fig.align="center", fig.width=12, fig.height=8}
myMixedPrimers <- oligos(roi, 
                         designStrategyPrimer = "mixed", 
                         probe = FALSE)

plotData(myMixedPrimers)
```

Results (first ten rows):

```{r, echo=FALSE}
myMixedPrimers[1:10, ] %>%
  kbl(., digits = 2) %>%
  kable_material(c("striped", "hover"), position = "right") %>%
  scroll_box(width = "650px", height = "300px")
```

Importantly, for mixed primers, identity refers to the mean identity of the target alignment at the 5' consensus part and coverage refers to the mean coverage of the target alignment at the 3' degenerate part. Values of 1 indicate perfect complementary to all target variants in the input alignment. 

**Masking**

For sequencing applications such as genotyping, there is often an interest in amplifying a specific region of the genome. In such cases, the primer dataset must be masked to not contain forward primers after the intended start position and to not contain reverse primers after the end position of the desired amplicon.

Below, I show how to create a subset of `myMixedPrimers`, where all primers between position 2000-3000 are removed: 

```{r}
myMixedPrimersMasked <- myMixedPrimers[
  (myMixedPrimers$fwd & myMixedPrimers$end < 2000) |
    (myMixedPrimers$rev & myMixedPrimers$start > 3000),
]

plotData(myMixedPrimersMasked)
```

### Step 3: `assays`

The next step is to find pairs of forward and reverse primers, and eventually, to combine them with probes. 

The arguments of `assays()` are:  

* `x`: the oligo dataset from Step 2
* `lengthRange`: amplicon length range, defaults to `c(65, 120)`
* `tmDiffPrimers`: maximum tm difference between the (mean tm) of the two primers (absolute value), defaults to `2`
* `tmDiffPrimersProbe`: acceptable tm difference between the primer pair and probe, defaults to `c(-10, 10)`. The tm difference is calculated by subtracting the mean tm of the probe with the mean tm of the primers. A positive value means that the average tm of the probe is higher than the average tm of the primer

The assays will contain probes if probes are present in the oligo dataset from Step 2. An error message will return if no assays are found.

Below, I show how to design assays using default settings using the dataset `myOligos` from Step 1.  

```{r}
myAssays <- assays(myOligos)  
```

Output (first ten rows): 

```{r, echo=FALSE}
myAssays[1:10, ] %>%
  kbl(., digits = 2) %>%
  kable_material(c("striped", "hover"), position = "right") %>%
  scroll_box(width = "650px", height = "300px")
```

The output contains many columns, so the best way of inspecting it is probably to type `View(as.data.frame(myAssays))` when using RStudio. 

Again, the results can be visualized using `plotData()`: 

```{r, fig.width=12, fig.height=6}
plotData(myAssays)
```

All assays are located within a conserved stretch of ~70 bases. This is also the target region for the most frequently used RT-qPCR assay for detection of HEV, developed by Jothikumar and colleagues [@jothikumar2006broadly].

There are now over 1000 assay candidates, but the number can be reduced by making a subset based on score. The assay score denotes the sum of the oligo score. Thus, the best possible score is 0 and the worst possible score is 24 for assays with only primers and 36 for assays with probes (the maximum possible score for each oligo is 12, see Step 2: `oligos`). 

```{r, fig.width=12, fig.height=6}
## Get the minimum score from myAssays
bestAssayScore <- min(myAssays$score)
bestAssayScore

## Make a subset that only contains the assays with the best scor 
myAssaySelection <- myAssays[myAssays$score == bestAssayScore, ]

## Plot these assays 
plotData(myAssaySelection)
```

**Design with modified settings**

To illustrate assay design with modified settings, I use the masked dataset from Step 2 (`masked`), where I designed primers to cover position 2000-3000. 

```{r}
myMixedAssays  <- assays(myMixedPrimersMasked, 
                         lengthRange = c(1000, 2000), 
                         tmDiffPrimers = 2)

plotData(myMixedAssays)
```

## Further handling of the data 

The oligo and assay region(s) can be inspected in more detail by using the consensus profile from Step 1. As an example, I select to take a closer look at the first assay in `myAssaySelection` from Step 3. 

The first step is to retrieve the start and end position of the assay, as follows:

```{r}
from <- myAssaySelection[1, ]$start
to <- myAssaySelection[1, ]$end
```

Now, it is possible to highlight the amplicon region, using the `highlight` argument in `plotData()`:

```{r fig 3, fig.width=12, fig.height=6}
plotData(myConsensusProfile, highlight = c(from, to))
```

To get a more detailed view, I can make a subset of the consensus profile that only contains the amplicon region: 

```{r}
myAssayRegion <- myConsensusProfile[
  myConsensusProfile$position >= from & 
    myConsensusProfile$position <= to, 
] 
```

And plot it:

```{r, fig.width=12, fig.height=6, fig.align="center"}
plotData(myAssayRegion)
plotData(myAssayRegion, type = "nucleotide")
```

The consensus amplicon sequence can be obtained by: 

```{r}
paste(myAssayRegion$iupac, collapse = "")
```

## Export to file

Before proceeding to wet-lab evaluation, it is highly recommended to also evaluate the primer and probe candidates for 1) the potential to form primer-dimers and hairpin structures, and 2), the potential to cross react with non-targets.

These tasks are better performed by other software. Below, I show how to export the results to file so that they can be analyzed by other tools.  

**Oligos and assays to fasta format**

`convertToDNAStringSet()` can be used for converting all sequence variants of all oligos or assays to a `Biostrings::DNAStringSet`-object. The `revAsRc` option regulates whether reverse primers should be written as reverse complement or not. 

```{r}
## Convert the first two oligos in myOligos 
convertToDNAStringSet(myOligos[1:2, ])

## Convert the first assay in myAssays 
convertToDNAStringSet(myAssays[1, ], revAsRc = FALSE)
```

The sequences can then be exported to fasta-format:

```r
toFile <- oligosToDNAStringSet(myOligos)
Biostrings::writeXStringSet(toFile, file = "myOligos.txt")

```

**Result tables**

Result tables can be saved to .txt or .csv-files, for instance: 

```r
write.csv(myAssays, file = "myAssays.csv", quote = FALSE, row.names = FALSE) 
write.table(myAssays, file = "myAssays.txt", quote = FALSE, row.names = FALSE) 
```

# Summary 

The design process is summarized below. 

```r
library(rprimer)

## Enter the filepath to an alignment with target sequences of interest 
filepath <- system.file("extdata", "example_alignment.txt", package = "rprimer")

## Import the alignment 
myAlignment <- Biostrings::readDNAMultipleAlignment(filepath, format = "fasta") 

## Design primers, probes and assays using default settings 
myConsensusProfile <- consensusProfile(myAlignment, ambiguityThreshold = 0.05)
myOligos <- oligos(myConsensusProfile)
myAssays <- assays(myOligos)

```

# Classes and example data

The "rprimer" classes (`RprimerProfile`, `RprimerOligo` and `RprimerAssay`) are extensions of the `DataFrame` class from S4Vectors [@PagesS4], and behave in a similar way as traditional data frames, with methods for `[`, `$`, `nrow`, `ncol`, `head`, `tail`, `rbind`, `cbind` etc. They can be coerced to traditional data frames by using `as.data.frame()`. 

Example datasets of each class are provided with the package, and are loaded by: 

```{r, warning=FALSE}
data("exampleRprimerProfile")
data("exampleRprimerOligo")
data("exampleRprimerAssay")
```

To provide reproducible examples, I've also included an alignment of class `Biostrings::DNAMultipleAlignment`. It is loaded by `data("exampleRprimerAlignment")`.

# Table values  

Lookup tables used for determination of complement bases, IUPAC consensus character codes, degeneracy and tm can be found by typing:

* `rprimer:::lookup$complement`
* `rprimer:::lookup$iupac`, `rprimer:::lookup$degenerates`, 
* `rprimer:::lookup$degeneracy` and 
* `rprimer:::lookup$nn`, respectively. 

# Source code 

The source code is available at https://github.com/sofpn/rprimer. 

# Citation

To cite rprimer, please use: `citation("rprimer")`.

# Session info 

```{r}
sessionInfo()
```

# References 
