---
title: "rprimer"
author: "Sofia"
package: rprimer
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Getting started}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

Primer and probe design are central parts of PCR assay development. Successful amplification and detection depend on several factors, such as optimal primer and probe length, GC-content, melting temperature and sequence complexity. Primer and probes must be designed so that formation of dimers and hairpin structures are avoided, and the intended target region must be sufficiently conserved to avoid severe mismatches between the primers/probe and template. This together makes PCR assay design a challenging task, which is not least reflected by the large number of computational methods available. One of the most comprehensive and widely used tools is Primer3. However, one important limitation with Primer3 is that the design process is based on a single sequence, which means that primers can be inadvertently placed in sequence heterogenous regions, resulting in amplification and quantification bias and poor coverage of the target sequences of interest (low inclusivity). This can, in part, be circumvented by making a “consensus profile”  from an alignment of the target sequences of interest, and masking the variable regions in the consensus sequence before the design process. However, for many applications, it can be difficult to find even a few completely conserved regions that are suitable as primer or probe targets. 

A common approach for amplifying sequence diverse targets is to use degenerate primers. A degenerate primer has one or more positions with several possible bases, and the degeneracy refers to the number of unique sequence variants it encompasses. These ambiguous bases are named according to a coding system determined by IUPAC. For instance, the IUPAC base K can be either G or T, and N can be A, C, G or T. For some applications, it is not uncommon to use primers with up to several hundreds or thousands of sequence variants. However, as the degeneracy increases, the likelihood of interaction with non-targets or other primer variants also rises, which risks to reduce both the sensitivity and specificity of the assay. 

Moreover, the proportional concentration of each sequence variants drops, 

An optimal design strategy would therefore be to find primer and probe binding regions with as low degeneracy as possible, but without compromising with the inclusivity. 

*ANOTHER PROBLEM IS THAT PRIMER CONC DROPS : https://academic.oup.com/nar/article/26/7/1628/1040055 : MM can be solved by reducing the annealing temp, but with the cost of less specific interactions XXXXXXXXX However, consensus primers may be too dissimilar to an unknown target to efficiently anneal to the original template, and degenerate primers may be too dissimilar to each other to efficiently amplify the synthesized product.*

If the degeneracy is to be kept low and the target sequence heterogeneous, it is next to impossible to fully avoid mismatches between the template and primers/probe. The effect of a mismatch depends to a great extent on its location. For primers, mismatches at the 3’ end can greatly reduce amplification efficiency and should best be avoided, while mismatches at the 5’ end are less detrimental. 


Mismatching sequences can successfully anneal if the annealing temperature is lowered. Or if primer length 
*[ADD CODEHOP INFO HERE]*

Mismatches within the binding region of a hydrolysis probe can be severe, especially at the 5' end, since the 5’-3’ exonuclease activity of the DNA polymerase relies on completely hybridized nucleotides to hydrolyse DNA duplexes. 

This R package ("rprimer") was developed to implement previously developed primer and probe design guidelines for sequence variable targets. It takes a multiple sequence alignment with target sequences of interest as input and designs primers, probes and PCR assays from user specified design criteria. 

In this document, I demonstrate how to use rprimer by designing a broadly-reactive RT-q/dPCR for detection of hepatitis E virus (HEV), which is a sequence variable RNA-virus and a common foodborne pathogen.  

# Installation

rprimer can be installed from [GitHub](https://github.com/) with:

``` r
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("sofpn/rprimer")
```
```{r, internal setup, echo=FALSE, warning=FALSE, message=FALSE}
library(kableExtra)
#devtools::load_all(".")
```
```{r setup, message=FALSE, warning=FALSE}
library(rprimer)
```

# Overview

The package contains four functions:

* `consensusProfile()`
* `oligos()`  
* `assays()` 
* `plotData()`

# Workflow

## Collection of target sequences and multiple alignment 

The first step is for the user to identify and align target sequences of interest. The file “example_alignment.txt” contains an alignment of 200 HEV sequences. The filepath is retrieved by:  

```r
system.file("extdata", "example_alignment.txt", package = "rprimer")
```

These sequences were collected by searching for “[a-z]”[porgn:__txid1678143] (the taxid of Ortohepevirus A, i.e., HEV) on the nucleotide database of NCBI GenBank, and filtering for sequence lengths > 6000, to retrieve complete and nearly complete genomes. The search had 674 hits (as of January 19, 2020). All sequences were downloaded (as complete record, fasta-file) and run through Hepatitis E virus genotyping tool version 0.1 (https://www.rivm.nl/mpf/typingtool/hev/, accessed January 19, 2020) to verify that they were correctly classified as Ortohepevirus A. Sequences that were not classified as Ortohepevirus A were removed, and the remaining ones (n = 642) were reduced to 200 by random selection, to not provide a too large example dataset for the package. The sequences were then aligned using the online version of mafft version 7 (https://mafft.cbrc.jp/alignment/server/, accessed January 19, 2020), using Auto settings.

## Data import 

Before the design process can begin, the user must import the alignment to R using the `readDNAMultipleAlignment()` function from the Biostrings-package. The input file can contain up to several thousands of sequences.

For example:

```{r}
infile <- system.file("extdata", "example_alignment.txt", package = "rprimer")

myAlignment <- Biostrings::readDNAMultipleAlignment(infile, format = "fasta")
```

For brevity, and to get the alignment to agree more with the average HEV genome length, I decided to mask all positions with a gap frequency of more than 50 %:   

```{r}
myAlignment <- Biostrings::maskGaps(myAlignment, 
                                    min.fraction = 0.5, 
                                    min.block.width = 1) 
```

However, this should be done with caution, since it may lead to placement of primers and probes in insert-rich regions. In such cases, a good idea could be to verify the design process without gap mask. 

## Design process

### Step 1: `consensusProfile` 

Once the alignment is imported, the first step is to generate a consensus profile. At each position in the alignment, the following information is retrieved:

* The proportion of the bases A, C, G, T, and “other” 
* The proportion of gaps (-)
* Majority consensus base, which is the most frequently occurring base 
* Identity, which is the proportion of the majority consensus base when gaps and ambiguous bases ("other") are excluded
* IUPAC consensus character (ambiguous base). The user can specify an “ambiguity threshold” between 0 and 0.2. All bases that occur with a proportion higher than that will be included in the ambiguous base. A threshold of 0 (default) will catch all the variation, but with the potential cost of generating primers and probes with high degeneracy, and a higher threshold will catch most of the variability, but with the potential cost of missing minor sequence variations (this information is provided by the coverage, see below)
* Shannon entropy, which is a measurement of variability. A value of zero indicate complete conservation, and a high value indicate high variability. Gaps and ambiguous bases are not included in this calculation 
* Coverage, which describes the total proportion of bases that are "covered" by the IUPAC consensus character. Gaps and ambiguous bases are not included in this calculation

Below, I make a consensus profile with an ambiguity threshold of 5 %:

```{r}
myConsensusProfile <- consensusProfile(myAlignment, ambiguityThreshold = 0.05)
```

Results (first ten rows):

```{r, echo=FALSE}
myConsensusProfile[1:10, ] %>%
  kbl(., digits = 2) %>%
  kable_material(c("striped", "hover"), position = "right") %>%
  scroll_box(width = "650px", height = "300px")
```

The results can be visualized with `plotData()`. 

```{r, fig.width=12, fig.height=6}
plotData(myConsensusProfile)
```

The dots represent the value at each position and the black lines represent centered running averages. High identity and low entropy values indicate high sequence conservation.

It is also possible to zoom into a specific region of interest:

```{r, fig.width=12, fig.height=6}
selection <- myConsensusProfile[
  myConsensusProfile$position >= 1000 & myConsensusProfile$position <= 1500, 
  ]

plotData(selection)
```

The nucleotide distribution can also be visualized, preferably on a short range (1-50 bases), by specifying `type = nucleotide`: 

```{r, fig.width=12, fig.height=6}
ntSelection <- myConsensusProfile[
  myConsensusProfile$position >= 1000 & myConsensusProfile$position <= 1030, 
  ]

plotData(ntSelection, type = "nucleotide")
```

### Step 2: `oligos`

The next step is to design oligos. Primers are mandatory to design and probes are optional. Primers can be designed in either one of two ways: "ambiguous" (default) or "mixed". Probes are always designed using the ambiguous strategy. 

* The **ambiguous strategy** generates primers and probes from the IUPAC consensus sequence. This means that degenerate bases can occur at any position in the oligo

* The **mixed strategy** generates primers from both the majority (~2/3 of the primer, 5' end) and IUPAC consensus sequence (~1/3 of the primer, 3' end). 

This means that the 5' end will lack ambiguous bases, and instead contain the most probable nucleotide at each position. This is a way of reducing the degeneracy for very variable targets, since 5' end mismatches are considered to be less detrimental than 3' end mismatches. This approach is (roughly) based on the Consensus-Degenerate Hybrid Oligonucleotide Primer (CODEHOP) principle

Recommended with long lengths... 

**Design settings**

The design process in `oligos()` can be modified in the following ways:  

* `maxGapFrequency`: maximum gap frequency (in the alignment) for primers and probes, defaults to `0.05`
* `lengthPrimer`: primer length, defaults to `18:22`
* `maxDegeneracyPrimer`: maximum degeneracy for primers, defaults to `4`
* `gcClampPrimer`: if primers should have a GC clamp, defaults to `TRUE`'. A GC clamp is here defined as the presence of two to three G or C:s within the last five bases (3' end) of the oligo. It is recommended for primers, since it helps to promote specific binding of the 3' end
* `avoidThreeEndRunsPrimer`: if primers with more than two runs of the same nucleotide at the terminal 3'-end should be avoided (to reduce the risk of mispriming), defaults to `TRUE`
* `minThreeEndCoveragePrimer`: minimum allowed coverage at the 3' end (the last five bases). If set to `1`, all bases of the 3' end must cover all sequence variants in the target alignment. Defaults to `0.98` 
* `gcRangePrimer`: GC-content range of primers (proportion), defaults to `c(0.40, 0.60)`
* `tmRangePrimer`: melting temperature range for primers, defaults to `c(55, 65)`. Melting temperature is calculated for perfectly matching oligo-target duplexes using the nearest neighbor method developed SantaLucia et al (REF). Formula, salt correction method and table values are from (REF). The table values can also be found by typing `rprimer:::lookup$nn`
* `concPrimer`: primer concentration in nM (for tm calculation), defaults to `500`
* `designStrategyPrimer`: design strategy for primers, `"ambiguous"` or `"mixed"`, defaults to `"ambiguous"`  
* `probe`: if probes should be designed as well, defaults to `TRUE`
* `lengthProbe`: defaults to `18:22`
* `maxDegeneracyProbe`: defaults to `4`
* `avoidFiveEndGProbe`: if hydrolysis probes with a G at the terminal 5' end should be avoided (to prevent quenching of the 5' florouphore), defaults to `TRUE` 
* `gcRangeProbe`: defaults to `c(0.40, 0.50)`
* `tmRangeProbe`: defaults to `c(55, 65)`
* `concProbe`: defaults to `250`
* `concNa` Sodium ion concentration (in M) in the PCR reaction (for tm calculation), defaults to `0.05` (50 mM)

All sequence variants must fall within the acceptance criteria for an oligo to be considered as valid. Moreover, oligos with sequence variants that contain more than four consecutive repeats of the same nucleotide (e.g. AAAAA) or more than three consecutive repeats of the same di-nucleotide (e.g. "ATATATAT") will be excluded. 

An error message will return if no oligos are found. If so, a good idea could be to re-run the process from Step 1 and increase the `ambiguityThreshold` in `consensusProfile()`, and/or relax the design constraints above. 

**Examples**

I start by designing primers and probes using default settings. 

``` {r}
myOligos <- oligos(myConsensusProfile)
```

Results (first ten rows):

```{r, echo=FALSE}
myOligos[1:10, ] %>%
  kbl(., digits = 2) %>%
  kable_material(c("striped", "hover"), position = "right") %>%
  scroll_box(width = "650px", height = "300px")
```

The man page (`?rprimer::oligos`) contains a detailed description of each column. However, what may not be completely self-explanatory is that some columns themselves contain vectors. For example: 

```{r}
myOligos$iupacSequence[1] # The IUPAC sequence of the first oligo 
myOligos$sequence[[1]] # All sequence variants of the first oligo 
```

* gc give the GC-content and melting temperature of all sequence variants: 
```{r}
myOligos$gcContent[[1]] # GC content of all variants of the first oligo 
myOligos$tm[[1]] # Tm of all variants of the first oligo 
```

All primer and probe candidates can be visualized using `plotData()`:

```{r, fig.align="center", fig.width=12, fig.height=8}
plotData(myOligos)
```

The upper plot shows the binding sites of the primers and probes (some are overlapped), and the lower plots show the features of the generated oligos. 

I now increase the maximum allowed degeneracy for primers, and lower the maximum allowed degeneracy for probes: 

```{r, fig.align="center", fig.width=12, fig.height=8}
myOligos2 <- oligos(myConsensusProfile, maxDegeneracyPrimer = 32, maxDegeneracyProbe = 1)

plotData(myOligos2)
```

I now use the mixed strategy, and design primers only, and modify some of the settings: 

```{r, fig.align="center", fig.width=12, fig.height=8}
myOligos4 <- oligos(myConsensusProfile, 
                    designStrategyPrimer = "mixed", 
                    maxDegeneracyPrimer = 8, 
                    tmRangePrimer = c(55, 75), 
                    minThreeEndCoverage = 0.99, 
                    lengthPrimer = 25:30, 
                    probe = FALSE)

plotData(myOligos4)
```

Results (first ten rows):

```{r, echo=FALSE}
myOligos4[1:10, ] %>%
  kbl(., digits = 2) %>%
  kable_material(c("striped", "hover"), position = "right") %>%
  scroll_box(width = "650px", height = "300px")
```

**Subsetting**

It is also possible to design oligos within a specific region of interest, by subsetting the `RprimerProfile` object:  

``` {r}
#myRoi <- myConsensusProfile[myConsensusProfile$position >= 5000 & myConsensusProfile$position <= #6000, ]
#myOligos3 <- oligos(myRoi)
```

```{r, fig.align="center", fig.width=12, fig.height=8}
#plotData(myOligos3)
```

### Step 3: `assays`

`assays()` finds pairs of forward and reverse primers within an `RprimerOligo`-object, and combines them with probes, if available. It returns an `RprimerAssay`-object.   

The design process is based on the following criteria:

* `lengthRange` Amplicon length range, defaults to `c(65, 120)`
* `maxTmDiffPrimers` Maximum Tm difference between the two primers (absolute value, calculated for majority primers), defaults to `2`
* `tmDiffPrimersProbe` Acceptable Tm difference between the primers (average Tm of the primer pair) and probe, defaults to `c(0, 20)`. The Tm-difference is calculated by subtracting the Tm of the (majority) probe with the average Tm of the (majority) primer pair. Hence, a negative Tm-difference means that the Tm of the probe is lower than the average Tm of the primer pair

```{r}
myAssays <- assays(myOligos)
```

Output (first ten rows): 

```{r, echo=FALSE}
myAssays[1:10, ] %>%
  kbl(., digits = 2) %>%
  kable_material(c("striped", "hover"), position = "right") %>%
  scroll_box(width = "650px", height = "300px")
```

The output contains many columns, so the most sensible way of inspecting it is probably to use `View(as.data.frame(myAssays))`, when using RStudio. 

Again, the output can be visualized using `plotData()`: 

```{r, fig.width=12, fig.height=6}
plotData(myAssays)
```

I notice that all assays are located in the conserved region between position 5000-5500. 

An error message will return if no assays are found.

## Further handling of the data 

The oligo and assay region(s) can be inspected further by using the `RprimerProfile` object. As an example, I select to take a closer look at the first assay in `myAssays`: 

```{r}

## Find start and end positions for the first assay 
from <- myAssays[1, ]$start
to <- myAssays[1, ]$end

## Subset the RprimerProfile object based on these positions
selection <- myConsensusProfile[myConsensusProfile$position >= from & myConsensusProfile$position <= to, ] 
```

```{r, fig.width=12, fig.height=6}
plotData(selection)
plotData(selection, type = "nucleotide")
```

The consensus amplicon sequence can be obtained by: 

```{r}
paste(selection$iupac, collapse = "")
```

It is also possible to highlight a specific region, using the optional argument `highlight`.

```{r fig 3, fig.width=12, fig.height=6}
plotData(myConsensusProfile, highlight = c(from, to))
```

## Save results to file  

The results can be saved to .txt or .csv-files, for instance: 

```r
write.csv(myAssays, file = "myAssays.csv", quote = FALSE, row.names = FALSE) 
write.table(myAssays, file = "myAssays.txt", quote = FALSE, row.names = FALSE) 
```

# Summary 

The design process is summarized below. In this case, I use default settings for oligo and assay design. 

```r
## Enter the filepath to an alignment with target sequences of interest 
filepath <- system.file("extdata", "example_alignment.txt", package = "rprimer")

## Import the alignment 
myAlignment <- Biostrings::readDNAMultipleAlignment(filepath, format = "fasta") 
myAlignment <- Biostrings::maskGaps(myAlignment, min.fraction = 0.5, min.block.width = 1)

## Design primers, probes and assays
myConsensusProfile <- consensusProfile(myAlignment, ambiguityThreshold = 0.05)
myOligos <- oligos(myConsensusProfile)
myAssays <- assays(myOligos)

```

# Classes

The "rprimer" classes (`RprimerProfile`, `RprimerOligo` and `RprimerAssay`) are extensions of the `DataFrame` class from S4Vectors, and behave in a similar manner as traditional data frames, with methods for `[`, `$`, `nrow`, `ncol`, `head`, `tail`, `rbind`, `cbind` etc. They can be coerced to traditional data frames by using `as.data.frame()`. 

Example data sets of each class are provided with the package, and are loaded by: 

```{r, warning=FALSE}
data("exampleRprimerProfile")
data("exampleRprimerOligo")
data("exampleRprimerAssay")
```

To provide reproducible examples, I've also included an alignment of class `Biostrings::DNAMultipleAlignment`. It is loaded by `data("exampleRprimerAlignment")`.

# Session info 

```{r}
sessionInfo()
```

# References 

Allawi, H. & SantaLucia, J. (1997) Thermodynamics and NMR of Internal G-T Mismatches in DNA. Biochemistry, 36, 34: 1058-110594

Pagès, H., Aboyoun, P., Gentleman, R., & DebRoy, S. (2020). Biostrings: Efficient manipulation of biological strings. R package version 2.57.2

Pagès, H., Lawrence M., & Aboyoun, P. (2020). S4Vectors: Foundation of vector-like and list-like containers in Bioconductor. R package version 0.27.12

SantaLucia, J., Hatim, T., Allawi H., & Seneviratne, A. (1996) Improved Nearest-Neighbor Parameters for Predicting DNA Duplex Stability. Biochemistry, 35: 3555-3562

SantaLucia, J. (1998) A unified view of polymer, dumbell, and oligonucleotide DNA nearest-neighbor thermodynamics. Proc. Natl. Acad. Sci. USA, 95: 1460-1465
