---
title: "rprimer"
author: "Sofia Persson"
package: rprimer
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Getting started with rprimer}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction 

rprimer provides tools for designing broadly reactive (RT)-(q/dd)PCR assays from multiple DNA sequence alignments. The design process is built on three functions: 

* `getConsensusProfile()`: returns an `RprimerProfile` object, which is used as input for;
* `getOligos()`: returns an `RprimerOligo` object, which is used as input for; 
* `getAssays()`: returns an `RprimerAssay` object

The `Rprimer`-classes are extensions of the `DataFrame` class from S4Vectors, and behave in a similar manner as traditional data frames (with methods for e.g. [, head, tail), and can be coerced to such by using `as.data.frame()`.

# Installation

rprimer can be installed from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("sofpn/rprimer")
```

Setup for the code in this document: 

```{r setup, message = FALSE, warning = FALSE}
library(rprimer)
library(magrittr)
library(Biostrings)
#devtools::load_all(".")
```

# Workflow 

## Import data 

The first step is to import an alignment with target sequences of interest and, if preferred, mask positions with high gap frequency. `readDNAMultipleAlignment()` and `maskGaps()` from Biostrings do the work for this part. 

Here, I  want to design an RT-(q/d)PCR assay for detection of hepatitis E virus, which is a highly variable RNA virus. The file "example_alignment.txt" is provided with the package and contains 100 hepatitis E virus sequences.  

I mask all positions with at least 50 % gaps. But note that masking positions with high gap frequency should we done with caution, since it may lead to that oligos are placed in positions were some sequences have inserts **(should I add an  avoid mask option in get oligos?)**

```{r}
infile <- system.file('extdata', 'example_alignment.txt', package = 'rprimer')

myAlignment <- infile %>%
  Biostrings::readDNAMultipleAlignment(., format = "fasta") %>%
  Biostrings::maskGaps(., min.fraction = 0.5, min.block.width = 1) 
```

## Step 1: `getConsensusProfile` 

`getConsensusProfile()` takes a `Biostrings::DNAMultipleAlignment` object as input and returns all the information needed for the subsequent design process. Note that masked positions are excluded. 

```{r}
myConsensusProfile <- getConsensusProfile(myAlignment, iupacThreshold = 0.05)
```

Output: 

```{r, echo=FALSE}
knitr::kable(head(myConsensusProfile), digits = 2)
```

Some comments on the data: 

* Majority refers to the majority consensus sequence, which is the most frequently occurring base. Identity is the proportion of that base when gaps and other bases than A, C, G and T are not taken into account

* The IUPAC consensus sequence includes wobble bases according to the IUPAC-nomenclature. It includes all DNA bases (A, C, G, T) that occurs with a frequency higher than the stated `iupacThreshold`. In my example, I set the threshold to `0.05`, which means that all bases that occur in more than 5 % of the sequences will be accounted for when the IUPAC base will be determined. 

* Entropy refers to Shannon entropy, which is a measurement of variability. A value of zero indicate no variability and a high value indicate high variability. Gaps and other bases than A, C, G and T are not included in this calculation 

The data can be visualized with `plotData()`, and specific regions can be highlighted using the optional arguments `shadeFrom` and `shadeTo`. The most conserved region of the hepatitis E virus genome appears to be between position 5000-5500:

```{r, fig.align="center", fig.width=12, fig.height=6}
plotData(myConsensusProfile, shadeFrom = 5000, shadeTo = 5500)
```

The black lines represent centered running averages and the blue dots represent the value at each position. 

## Step 2: `getOligos`

`getOligos()` searches for oligos from an `RprimerProfile`-object. All oligos are shown in both majority (without degenerate bases) and IUPAC format (with degenerate bases). The oligos are designed from the following constraints:  

* `lengthPrimer` Primer length, defaults to `18:22`
* `maxGapFrequencyPrimer` Maximum gap frequency for primers, defaults to `0.1`
* `maxDegeneracyPrimer` Maximum number of variants of each primer, defaults to `4`
* `gcClampPrimer` If primers must have a GC-clamp, defaults to `TRUE`
* `avoid3endRunsPrimer` If primers with more than two runs of the same nucleotide at the terminal 3' end should be avoided, defaults to `TRUE`
* `minEndIdentityPrimer` Optional. Minimum allowed identity at the 3' end (i.e. the last five bases). E.g., if set to `1`, only primers with complete target conservation at the 3' end will be considered
* `gcRangePrimer` GC-content-range for primers, defaults to `c(0.45, 0.55)`
* `tmRangePrimer` Melting temperature (Tm) range for primers, defaults to `c(55, 65)`. Tm is calculated using the nearest-neighbor method. See `?rprimer::getOligos` for a detailed description and references
* `concPrimer` Primer concentration in nM (for Tm calculation), defaults to `500`
* `probe` If probes should be designed as well, defaults to `TRUE`
* `lengthProbe` Probe length, defaults to `18:22`
* `maxGapFrequencyProbe` Maximum gap frequency for probes, defaults to `0.1` 
* `maxDegeneracyProbe` Maximum number of variants of each probe, defaults to `4`
* `avoid5endGProbe` If probes with a G at the terminal 5' end should be avoided, defaults to `TRUE`
* `gcRangeProbe` GC-content-range for probes, defaults to `c(0.45, 0.55)`
* `tmRangeProbe` Melting temperature (Tm) range for probes, defaults to `c(55, 70)`
* `concProbe` Probe concentration in nM (for Tm calculation), defaults to `250`
* `concNa` Sodium ion concentration in the PCR reaction (for Tm calculation), defaults to `0.05` M (50 mM)
* `showAllVariants` If sequence, GC-content and Tm should be presented for all variants of each oligo (in case of degenerate bases).`TRUE` (slower) or `FALSE` (faster), defaults to `TRUE`

In addition, `getOligos()` avoids:

* Majority oligos with more than than three consecutive runs of the same dinucleotide (e.g. "TATATATA")
* Majority oligos with more than four consecutive runs of the same nucleotide  (e.g. "AAAAA")
* Majority oligos that are duplicated (to prevent binding at several places on the genome)

An error message will return if no oligos are found. 

Below, I design both primers and probes for my hepatitis E virus target. I want an assay that targets a conserved region as possible. For this reason, I use very strict constraints on degeneracy (not more than 2) and minimum end identity (1) for primers.

``` {r}
myOligos <- getOligos(myConsensusProfile,
                      lengthPrimer = 18:22,
                      maxGapFrequencyPrimer = 0.05,
                      maxDegeneracyPrimer = 2,
                      gcClampPrimer = TRUE,
                      avoid3EndRunsPrimer = TRUE,
                      minEndIdentityPrimer = 1,
                      gcRangePrimer = c(0.45, 0.65),
                      tmRangePrimer = c(55, 65),
                      concPrimer = 500,
                      probe = TRUE,
                      lengthProbe = 16:24,
                      maxGapFrequencyProbe = 0.05,
                      maxDegeneracyProbe = 4,
                      avoid5EndGProbe = TRUE, 
                      gcRangeProbe = c(0.45, 0.65),
                      tmRangeProbe = c(55, 70),
                      concProbe = 250,
                      concNa = 0.05,
                      showAllVariants = TRUE)
```

Some comments on the output:

* NA means that the oligo was invalid, due to e.g. missing GC-clamp or an end-identity below the threshold

All oligo candidates can be visualized using `plotData()`:

```{r, fig.align="center", fig.width=12, fig.height=8}
plotData(myOligos)
```

## Step 3: `getAssays`

`getAssays()` finds pairs of forward and reverse primers and combines them with probes (if selected). The design process is based on the following constraints: 

* `length` Amplicon length, defaults to `65:120`
* `maxTmDifferencePrimers` Maximum Tm difference between the two primers (absolute value, calculated for majority primers), defaults to `2`
* `tmDifferencePrimersProbe` Acceptable Tm difference between the primers (average Tm of the primer pair) and probe, defaults to `c(0, 20)`. The Tm-difference is calculated by subtracting the Tm of the (majority) probe with the average Tm of the (majority) primer pair. A negative Tm-difference means that the Tm of the probe is lower than the average Tm of the primer pair

An error message will return if no assays are found.

```{r}
myAssays <- getAssays(myOligos, 
                      length = 65:120,
                      maxTmDifferencePrimers = 2,
                      tmDifferencePrimersProbe = c(-2, 10))
```

Some comments on the output: 

* Probes can be either positive or negative sense. If both the positive and negative sense probe are considered valid for a specific assay, the probe with the least number of Gs will be selected. **More!**

The output can be visualized using `plotData()`: 

```{r, fig.align="center", fig.width=12, fig.height=6}
plotData(myAssays)
```
All assays are located in the conserved region between position 5000-5500. 

The `RprimerProfile` object can be used to further inspect the assay region(s). I select to investigate the first assay in `myAssays`: 

```{r}
assayStart <- myAssays[1, ]$start
assayEnd <- myAssays[1, ]$end
myAssayRegion <- myConsensusProfile[myConsensusProfile$position >= assayStart & myConsensusProfile$position <= assayEnd, ] 
```

The nucleotide distribution within the amplicon can be visualized with `plotNucleotides()`. The `rc` option regulates whether the sequence should be displayed as a reverse complement or not.

```{r, fig.align="center", fig.width=12, fig.height=4}
plotNucleotides(myAssayRegion, rc = FALSE) 
```

I can of course also use `plotData`:

```{r, fig.align="center", fig.width=12, fig.height=6}
plotData(myAssayRegion)
```

The amplicon sequence can be obtained by: 

```{r}
paste(myAssayRegion$majority, collapse = "")
```

## Summary 

The design process is summarized in the following pipeline. In this case, I use default settings for oligo and assay design. I save all the different objects so I can plot them, or write them to a .txt or .csv file. 

```{r}
## Enter the filepath to the alignment here 
filepath <- system.file('extdata', 'example_alignment.txt', package = 'rprimer')

myAssays <- filepath %>%
  Biostrings::readDNAMultipleAlignment(., format = "fasta") %>%
  {Biostrings::maskGaps(
    ., min.fraction = 0.5, min.block.width = 1
  ) ->> myAlignment} %>%
  {getConsensusProfile(., iupacThreshold = 0.05) ->> myConsensusProfile} %>%
  {getOligos(.) ->> myOligos} %>%
  getAssays(.)
  
# plotData(myConsensusProfile)
# plotData(myOligos)
# plotData(myAssays)
```

# Example data 

Example data sets of each class are provided with the package, and can be loaded by: 

```{r}
data("exampleRprimerAlignment")
data("exampleRprimerProfile")
data("exampleRprimerOligo")
data("exampleRprimerAssay")
```

# Session info 

```{r}
sessionInfo()
```
