---
title: "Getting started with rprimer"
author: "Sofia Persson"
package: rprimer
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Getting started with rprimer}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, message = FALSE, warning = FALSE}
library(rprimer)
library(magrittr)
#devtools::load_all(".")

```

## Introduction 

There are several factors to take into account when designing primers, 
probes and PCR-assays, for instance:  

* GC content
* Melting temperature
* Sequence conservation 
* ....

The purpose of rprimer is to automate primer, 
probe and PCR assay design as much as possible. 

In this document, I demonstrate how to use rprimer by designing a real-time 
RT-PCR assay for detection of hepatitis E virus (HEV), 
which is a highly variable RNA virus and a common foodborne pathogen. 

## Installation

Install rprimer by calling: 

```{r}
# if (!requireNamespace("devtools", quietly = TRUE)) {
#    install.packages("devtools")
# }
# devtools::install_github("sofpn/rprimer")

# library(rprimer)
```

### Data structures

* RprimerProfile
* RprimerProperties 
* RprimerOligo
* RprimerAssay 

### Worked examples 

### Step 1: Import alignment 
Already existing functionality can be used to import an alignment.

In this example, I use an alignment with 100 HEV sequences collected 
from GenBank, which is provided with the package. 

```{r}
#infile <- system.file('extdata', 'exampleAlignment.txt', package = "rprimer")
```

```{r}
# Import alignment and remove positions with high gap frequency
#myAlignment <- infile %>%
#  Biostrings::readDNAMultipleAlignment(., format = "fasta") %>%
#  Biostrings::maskGaps(., min.fraction = 0.5, min.block.width = 1)

# Select region of interest
# Biostrings::colmask(myAlignment, invert = TRUE) <- IRanges::IRanges(start = 1000, end = 5000)

```



### Session info

```{r}
sessionInfo()
```


